### DISCUSSION TITLE
What are the differences of -crop, -extract, and -region?

### DISCUSSION THREAD

**Discussion by alxtsg** [ImageMagick/ImageMagick#5386] [Help] [answered]
According to the documentation, ImageMagick (v7) supports the following options:

* [-crop](https://imagemagick.org/script/command-line-options.php#crop)
* [-extract](https://imagemagick.org/script/command-line-options.php#extract)
* [-region](https://imagemagick.org/script/command-line-options.php#region)

Assume I have 2 images of the same dimension (1920px by 1080px), color profile, format (JPEG). Now I want to compare the specific region of the 2 images, what are the differences of the following commands?

```
magick image.jpeg reference.jpeg -metric ssim -compare -crop 1920x1065+0+0 -format "%[distortion]" info:
magick image.jpeg reference.jpeg -metric ssim -compare -extract 1920x1065+0+0 -format "%[distortion]" info:
magick image.jpeg reference.jpeg -metric ssim -compare -region 1920x1065+0+0 -format "%[distortion]" info:
```

The comparison results are the same, and the time needed to complete the commands are roughly the same. So I'm wondering which one is more efficient.

**Comment by fmw42** [comment]
I believe that you need to put your -crop or -extract before -compare in order to crop the two input images. If the two images are the same and the region you want is smaller then -crop and -extract should be the same.  -region is deprecated I believe in IM 7. Or at least it does not work the same as in IM 6. It is not recommended.  -region processes only in that region. It does not crop the image. So the output if used would be the same size as the input.

**Reply by alxtsg** [reply]
Thanks for the quick reply.

> -region is deprecated I believe in IM 7.

The documentation doesn't mention that but okay let's focus on `-crop` and `-extract`.

I changed the order the `-compare` and `-crop`/ `-extract` options, the results are not the same.

```
magick image.jpeg reference.jpeg -metric ssim -crop 1920x1065+0+0 -compare -format "%[distortion]" info:
# Output: 0.995816

magick image.jpeg reference.jpeg -metric ssim -extract 1920x1065+0+0 -compare -format "%[distortion]" info:
# Output: 0.995702
```

I also tried the [`magick compare`](https://imagemagick.org/script/compare.php) sub-commands:

```
magick compare -metric ssim -crop 1920x1065+0+0 reference.jpeg image.jpeg - | Out-Null
magick compare -metric ssim -extract 1920x1065+0+0 reference.jpeg image.jpeg - | Out-Null
# Both output the same result: 0.995816
```

So I guess I should use `-crop`?

**Reply by fmw42** [reply]
it is not -extract, what I though you wanted was -extent, which should be the same as -crop in this case.

I have not uses -extract, but from the docs, it would appear that it should be the same -crop.  Not sure why you see any difference.

Also I would have put -crop and -extract right after reading the input images and before -metric skim.

**Reply by fmw42** [reply]
-extract needs to come before reading the input.  All the following are equivalent.

Input:

![wizard](https://user-images.githubusercontent.com/13458218/181621547-aa497985-0cf7-4330-ac2f-ad0ec3b3c815.jpg)

![wizard_blur1](https://user-images.githubusercontent.com/13458218/181621579-3bc46d03-a0f6-4aff-96a5-b575004ab823.jpg)



```
magick wizard.jpg wizard_blur1.jpg -crop 900x900+0+0 -metric ssim -compare -format "%[distortion]\n" info:
0.895627

magick wizard.jpg wizard_blur1.jpg -extent 900x900+0+0 -metric ssim -compare -format "%[distortion]\n" info:
0.895627

magick wizard.jpg[900x900+0+0] wizard_blur1.jpg[900x900+0+0] -metric ssim -compare -format "%[distortion]\n" info:
0.895627

magick -extract 900x900+0+0 wizard.jpg wizard_blur1.jpg -metric ssim -compare -format "%[distortion]\n" info:
0.895627


```

**Comment by snibgo** [comment]
As fmw42 says, you do the crop etc *after* you have done "`-compare`", so it has no effect on the comparison results. You need to put it *before*.

"`-extent`" is generally used to extend images, ie make them larger. It also flattens them.

"`-crop`" is used when you only want part of an image.

"`-region`" is entirely different.

**Reply by alxtsg** [reply]
Thanks for the reply.

> "-extent" is generally used to extend images, ie make them larger. It also flattens them.

As I only want to compare part of the images, I think this option isn't suitable for the task.

I did another experiment (as mentioned [above](https://github.com/ImageMagick/ImageMagick/discussions/5386#discussioncomment-3272064)) to compare the results with the `magick compare` sub-command, looks like using `-crop` option in IM v7 gives the same result as in IM v6, while `-extract` doesn't.

**Comment by snibgo** [comment, marked as answer]
Sorry, I misread your posts. You said "-extract", but I misread that as "-extent".

"`-extract`" is used *before* reading an image (from a file) to read just some of the pixels. "`-extract`" *after* reading the image does nothing. See https://imagemagick.org/script/command-line-options.php#extract . In your case, as far as I can see, it provides no benefit compared to "`-crop`" *after* reading the image.
